pkg/scheduler/queue/user_queues.go

```diff
@@ -28,298 +27,350 @@ type querier struct {
+	// Tenant querier ID is set to nil if sharding is off or available queriers <= tenant's maxQueriers.
+	tenantQuerierIDs map[string]map[string]struct{}
+}
+
+type queueTenant struct {
+	// seed for shuffle sharding of queriers; computed from userID only,
+	// and is therefore consistent between different frontends.
+	shuffleShardSeed int64
 
-	// Tracks queriers registered to the queue.
-	queriers map[string]*querier
+	// points up to user order to enable efficient removal
+	orderIndex int
 
-	// Sorted list of querier names, used when creating per-user shard.
-	sortedQueriers []string
+	maxQueriers int
 }
 
-type userQueue struct {
-	requests *list.List
+// This struct holds user queues for pending requests. It also keeps track of connected queriers,
+// and mapping between users and queriers.
+type queues struct {
+	tenantQueues map[string]*userQueue
 
-	// If not nil, only these queriers can handle user requests. If nil, all queriers can.
-	// We set this to nil if number of available queriers <= maxQueriers.
-	queriers    map[string]struct{}
-	maxQueriers int
+	tenantQuerierState tenantQuerierState
 
-	// Seed for shuffle sharding of queriers. This seed is based on userID only and is therefore consistent
-	// between different frontends.
-	seed int64
+	maxUserQueueSize int
+}
 
-	// Points back to 'users' field in queues. Enables quick cleanup.
-	index int
+type userQueue struct {
+	requests *list.List
 }
 
 func newUserQueues(maxUserQueueSize int, forgetDelay time.Duration) *queues {
 	return &queues{
-		userQueues:       map[string]*userQueue{},
-		users:            nil,
+		tenantQueues: map[string]*userQueue{},
+		tenantQuerierState: tenantQuerierState{
+			queriersByID:       map[string]*querierConn{},
+			querierIDsSorted:   nil,
+			querierForgetDelay: forgetDelay,
+			tenantIDOrder:      nil,
+			tenantsByID:        map[string]*queueTenant{},
+			tenantQuerierIDs:   map[string]map[string]struct{}{},
+		},
 		maxUserQueueSize: maxUserQueueSize,
-		forgetDelay:      forgetDelay,
-		queriers:         map[string]*querier{},
-		sortedQueriers:   nil,
 	}
 }
 
 func (q *queues) len() int {
-	return len(q.userQueues)
+	return len(q.tenantQueues)
 }
 
 func (q *queues) deleteQueue(userID string) {
-	uq := q.userQueues[userID]
+	uq := q.tenantQueues[userID]
 	if uq == nil {
 		return
 	}
+	u := q.tenantQuerierState.tenantsByID[userID]
 
-	delete(q.userQueues, userID)
-	q.users[uq.index] = ""
+	delete(q.tenantQueues, userID)
+	delete(q.tenantQuerierState.tenantsByID, userID)
+	q.tenantQuerierState.tenantIDOrder[u.orderIndex] = ""
 
 	// Shrink users list size if possible. This is safe, and no users will be skipped during iteration.
-	for ix := len(q.users) - 1; ix >= 0 && q.users[ix] == ""; ix-- {
-		q.users = q.users[:ix]
+	for ix := len(q.tenantQuerierState.tenantIDOrder) - 1; ix >= 0 && q.tenantQuerierState.tenantIDOrder[ix] == ""; ix-- {
+		q.tenantQuerierState.tenantIDOrder = q.tenantQuerierState.tenantIDOrder[:ix]
 	}
```

Comment:
Should this logic be pushed down into a method on `tenantQuerierState`?

Reply:
done